RFC-0004-jdbc-join-pushdown.md

```diff
@@ -0,0 +1,513 @@
+
+![joinpushdown_summary](https://github.com/Ajas-Mangal/jdbc-join-pushdown/assets/175085180/b0b4bddb-5ef9-40fc-a51f-d6c34ba36bdb)
+
+#### Join pushdowns for bushy trees & queries involving multiple connectors
+
+This jdbc join pushdown proposal is based on the existing presto flow and it is optimizing the PlanNode using JdbcJoinPushdown optimizer to pushing down the join node to connector level. 
+
+Presto always produces a left deep Join tree for a PlanNode (JoinNode). The JoinNode never met another JoinNode on right, and the right always will be a source for TableScanNode. But the left of the JoinNode could be another JoinNode. 
+
+That is, for a join query the left always resolve from 'FROM' clause table. So if we use a federated join query (which involves multiple connectors) we could pushdown all the tables of the connector which is defined as 'FROM' table. No other connector tables are going to pushing down as the left is always resolve from 'FROM' clause table and the right is the other connector level table. Below is the possible pushdown if we use postgres table as from table.
+
+
+![PlanNode_optimized](https://github.com/Ajas-Mangal/jdbc-join-pushdown/assets/175085180/77ff4b8c-3d25-4d1c-800c-d8885791407a)
+
+
+For doing federated join pushdown (pushing down tables from multiple connector) either we need to group the tables based on connector or need to rewrite the PlanNode to build right JoinNode instead of TableScanNode (bushy join). It is not planning on this proposal and here we are planning to pushing down all the tables from the single connector which is defined as 'FROM' clause.
+
+## Lowlevel Design
+
+#### Below is the detailed flow for jdbc join pushdown.
+
+![join_pushdown_detailed_design](https://github.com/Ajas-Mangal/jdbc-join-pushdown/assets/175085180/26c4ba30-28aa-47c5-991a-2abebac79ac1)
+
+#### Below are new points that Proposed for Join Pushdown Implementation
+
+* JoinNode make available on ConnectorPlanOptimizer
+
+* On ApplyConnectorOptimization Optimizer, add new ConnectorPlanOptimizer called JdbcJoinPushdown
+
+* JdbcJoinPushdown needs to implement ConnectorPlanOptimizer and the optimze method use ConnectorPlanRewriter to rewrite the JoinNode to a TableScanNode
+
+* Handle JoinNode on JdbcJoinPushdown ConnectorPlanRewriter. 
+
+* visitJoinNode to identify which Join nodes can be pushed down and handled by the connector. Multiple JoinNodes need to handle here.
+
+* If the connector, identifies a JoinNode that it wants to rewrite, collapsing the JoinNode into a new TableScanNode property by removing the JoinNode.
```

Comment:
For the presto, for a join query, the first case only available with deep left join. Its left and right are TableScanNode and this table details will add to JdbcTableHandle using the list `joinTables` . Now we create a new TableScanNode for this deep left JoinNode with above created JdbcTableHandle and replace that joinnode with this TableScanNode. Now the second case arrived as previous deep join was left child of its parent JoinNode and this parent now having replaced tablescannode as left. So it matches with its right table and add this right table too to that newly created JdbcTableHandle list `joinTables` and replace this JoinNode. This will continue bottom to top until mismatch join pushdown condition.